### PR TITLE
Fixes #2966 - pgr_withPoints does not pick optimal route when fraction=1

### LINKED ISSUE #2966: pgr_withPoints does not pick optimal route
**Problem**
pgr_withPoints does not pick the optimal route if the percentage of the (in my case target) point is 1. A percentage value of 0.99 or similar yields the appropriate results. 

**To Reproduce**

Minimal test case: 
* Three links: 100 = start, 101 and 102 are parallel with different costs (102 having the much higher cost)
* We want to go to from the middle of our start link to the end of 101 (i.e. node 1003).

```
SELECT * FROM pgr_withPoints(
  'SELECT * FROM (VALUES 
    (100, 1001, 1002, 10, 10),
    (101, 1002, 1003, 12, 12),
    (102, 1002, 1003, 100, 100))
  AS t(id, source, target, cost, reverse_cost)',
  'SELECT * FROM (VALUES
    (1, 100, 0.5),
    (2, 101, 1))
  AS t(pid, edge_id, fraction)',
  -1,
  -2
)
```

**Expectation**
Routing via 100 and 101

```
1	1	-1	100	5	0
2	2	1002	101	12	5
3	3	-2	-1	0	17
```

**Actual**
Routing goes via link 102, which is much more expensive.
```
1	1	-1	100	5	0
2	2	1002	102	100	5
3	3	1003	101	0	105
4	4	-2	-1	0	105
```

**Platform/versions**

```
SELECT version();
> "PostgreSQL 14.13 (Debian 14.13-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit"
SELECT postgis_full_version();
> "POSTGIS=""3.5.2 dea6d0a"" [EXTENSION] PGSQL=""140"" GEOS=""3.9.0-CAPI-1.16.2"" PROJ=""7.2.1 NETWORK_ENABLED=OFF URL_ENDPOINT=https://cdn.proj.org USER_WRITABLE_DIRECTORY=/var/lib/postgresql/.local/share/proj DATABASE_PATH=/usr/share/proj/proj.db"" (compiled against PROJ 7.2.1) LIBXML=""2.9.10"" LIBJSON=""0.15"" LIBPROTOBUF=""1.3.3"" WAGYU=""0.5.0 (Internal)"""
SELECT pgr_version();
> "3.8.0"
```

### PR DESCRIPTION
**Pull request by Mohit242-bit**
Fixes #2966

## Problem
Previously, When a point is placed at fraction=1 (exactly at the end of an edge), pgr_withPoints fails to find the optimal route.

**Expected Behavior:** Route via 100 -> 101 (Cheap) -> Cost 17
**Actual Buggy Behavior:** Route via 100 -> 102 (Expensive detour) -> Cost 105

**Root Cause**
->When fraction=1, the code only created a zero-cost link between the target node and the point. It did not create the   traversable edge from the source to the point making the direct path invisible to Dijkstra.

**Solution:**
- Updated the logic in `src/withPoints/withPoints.cpp` to create the correct edge for the point at fraction=1, allowing Dijkstra to find the optimal path.
- Added a regression test: `pgtap/withPoints/withPoints/edge_cases/issue_2966.pg` to ensure this bug does not return.

Now, when fraction=1:
-> Zero-cost edge from target node to point (same location).[ unchanged ]
-> **New**: Costed edge from source to point with proper traversal cost

## Testing
All tests pass:
bugfix

@pgRouting/admins

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixes route selection when points lie at edge ends (fraction = 1), ensuring correct per-segment cost accounting and optimal path choice.

* **Tests**
  * Adds regression pgTAP test suites covering withPoints, withPointsVia, trsp_withPoints and trspVia_withPoints edge cases; tests are version-gated (requires 4.1.0+) and run rollback-safe.

* **Documentation**
  * Updates release notes, translation strings, and contributor listings to document the fix and release acknowledgments.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/withPoints/withPoints.cpp [hunk 1]
```diff
@@ -397,8 +397,52 @@ Pg_points_graph::create_new_edges() {
...
+                        double cost_to_point = delta * edge.cost;
+                        Edge_t cost_edge = {
+                            edge.id,
+                            prev_target,
+                            point.vertex_id,
+                            cost_to_point,
+                            -1};
+                        m_new_edges.push_back(cost_edge);
+                        last_cost = cost_to_point;
+                        log << "Issue2966 fix: new_edge("
```
Comment (cvvergara) [inline src/withPoints/withPoints.cpp]: Please remove this log statement

File: src/withPoints/withPoints.cpp [hunk 2]
```diff
@@ -397,8 +397,52 @@ Pg_points_graph::create_new_edges() {
...
+                            << "id, source, target, cost, reverse_cost) = ("
+                            << cost_edge.id << "\t"
+                            << cost_edge.source << "\t"
+                            << cost_edge.target << "\t"
+                            << cost_edge.cost << "\t"
+                            << cost_edge.reverse_cost << ")\n";
+                    }
+                    if (edge.reverse_cost >= 0) {
+                        double rcost_to_point = delta * edge.reverse_cost;
+                        Edge_t rcost_edge = {
```
Comment (cvvergara) [inline src/withPoints/withPoints.cpp]: Can you put this in one line please, we have 120 characters allowed on C++ on our code

File: src/withPoints/withPoints.cpp [hunk 3]
```diff
@@ -397,8 +397,52 @@ Pg_points_graph::create_new_edges() {
...
+                Edge_t zero_edge = {edge.id, edge.target, -point.pid, 0, 0};
+                m_new_edges.push_back(zero_edge);
+
+                // Also create edge from previous target to this point
+                // This allows Dijkstra to find the direct path through this edge
+                double delta = point.fraction - prev_fraction;
+                if (delta > 0) {
+                    if (edge.cost >= 0) {
+                        double cost_to_point = delta * edge.cost;
+                        Edge_t cost_edge = {
```
Comment (cvvergara) [inline src/withPoints/withPoints.cpp]: Can you put this in one line please, we have 120 characters allowed on C++ on our code

File: src/withPoints/withPoints.cpp (hunk 1 above)
Comment (Mohit242-bit) [reply src/withPoints/withPoints.cpp]: yes done in recent push

File: src/withPoints/withPoints.cpp (hunk 2 above)
Comment (Mohit242-bit) [reply src/withPoints/withPoints.cpp]: yes done

File: pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  END IF;
+
+  CREATE TEMP TABLE test_edges AS
+  SELECT * FROM (VALUES
+    (100::BIGINT, 1001::BIGINT, 1002::BIGINT, 10::FLOAT, 10::FLOAT),
+    (101::BIGINT, 1002::BIGINT, 1003::BIGINT, 12::FLOAT, 12::FLOAT),
+    (102::BIGINT, 1002::BIGINT, 1003::BIGINT, 100::FLOAT, 100::FLOAT)
+  ) AS t(id, source, target, cost, reverse_cost);
+
+  -- Test 1: fraction = 1 should work correctly
```
Comment (cvvergara) [inline pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg]: Remove these comments that you are adding before a test

File: pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    $$SELECT agg_cost::TEXT FROM pgr_trsp_withPoints(
+      'SELECT * FROM test_edges',
+      'SELECT path, cost FROM (SELECT ARRAY[1]::INTEGER[] as path, 0::FLOAT as cost WHERE false) AS t',
+      'SELECT * FROM (VALUES (1, 100, 0.5), (2, 101, 1)) AS t(pid, edge_id, fraction)',
+      -1, -2,
+      'b',
+      true
+    ) WHERE edge < 0$$,
+    ARRAY['17']::TEXT[],
+    'Issue 2966 (TRSP): fraction=1 should route correctly (cost 17, not 105)'
```
Comment (cvvergara) [inline pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg]: The comment is removed because this is the comment.
which by the way please rewrite to something like this:
pgr_trsp_withPoints: expected cost is 17

File: pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  );
+
+  -- Test 2: Verify the path uses the correct edges
```
Comment (cvvergara) [inline pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg]: delete comment

File: pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    $$SELECT edge::TEXT FROM pgr_trsp_withPoints(
+      'SELECT * FROM test_edges',
+      'SELECT path, cost FROM (SELECT ARRAY[1]::INTEGER[] as path, 0::FLOAT as cost WHERE false) AS t',
+      'SELECT * FROM (VALUES (1, 100, 0.5), (2, 101, 1)) AS t(pid, edge_id, fraction)',
+      -1, -2,
+      'b',
+      true
+    ) WHERE edge > 0$$,
+    ARRAY['100', '101']::TEXT[],
+    'Issue 2966 (TRSP): Path should use edges 100 and 101'
```
Comment (cvvergara) [inline pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg]: rewrite:
pgr_trsp_withPoints: Path should use edges 100 and 101

File: pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+SELECT CASE WHEN min_version('4.1.0') THEN plan(2) ELSE plan(1) END;
+
+CREATE OR REPLACE FUNCTION issue_2966_tests()
+RETURNS SETOF TEXT AS
+$BODY$
+BEGIN
+
+  IF NOT min_version('4.1.0') THEN
+    RETURN QUERY SELECT skip(1, 'pgr_trsp_withPoints: Issue 2966 test requires 4.1.0+');
```
Comment (cvvergara) [inline pgtap/trsp/trsp_withpoints/edge_cases/issue_2966.pg]: Rewrite:
` Issue 2966 test requires 4.1.0+` -> `Fix implemented on 4.1.0`

File: NEWS.md [hunk 9]
```diff
@@ -12,7 +12,9 @@ To read all history of releases go to the latest [release notes](https://docs.pg
 To see all issues & pull requests closed by this release see the
 [#4.1.0](https://github.com/pgRouting/pgrouting/issues?utf8=%E2%9C%93&q=milestone%3A%22Release%204.1.0%22)
 
-No changes yet
+Bug Fixes
+
+* [#2966](https://github.com/pgRouting/pgrouting/issues/2966): pgr_withPoints does not pick optimal route when fraction = 1
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply NEWS.md:17]: I don't know what you did to add so many blank lines in this file
For this File you need to run:
```
tools/release-scripts/notes2news.pl 
```
After that if you run:
```
git diff upstream/develop NEWS.md
```
You will get only this small change instead of the bunch of changes adding a balnk lines
```
diff --git a/NEWS.md b/NEWS.md
index e7806f69af..1be7654c95 100644
--- a/NEWS.md
+++ b/NEWS.md
@@ -12,7 +12,9 @@ To read all history of releases go to the latest [release notes](https://docs.pg
 To see all issues & pull requests closed by this release see the
 [#4.1.0](https://github.com/pgRouting/pgrouting/issues?utf8=%E2%9C%93&q=milestone%3A%22Release%204.1.0%22)
 
-No changes yet
+Bug Fixes
+
+* [#2966](https://github.com/pgRouting/pgrouting/issues/2966): pgr_withPoints does not pick optimal route when fraction = 1
```

**Review by cvvergara — COMMENTED** [review]
Looks like its working :-) thanks
Left you with some cleanup tasks on the code


Now, It is code that its used on many pgRouting functions not only on the pgr_withPoints
See this [documentation](https://docs.pgrouting.org/dev/en/withPoints-family.html)
and a couple more are [here](https://docs.pgrouting.org/dev/en/TRSP-family.html)

**Comment by cvvergara** [comment]
Hi @Mohit242-bit 
I left some example changes to do on the pgtap tests
Also there is a section on `doc/src/pgRouting-introduction.rst` that needs change

From:
```
Individuals in this release v4.0.0 (in alphabetical order)                      
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
                                                                                
Bipasha Gayary,                                                                 
Fan Wu,                                                                         
Regina Obe,                                                                     
Saloni kumari,                                                                  
Vicky Vergara
```

to & add your name:
```
Individuals in this release v4.0.0 (in alphabetical order)                      
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
                                                                                
Regina Obe,                                                                     
Vicky Vergara
```

also add your name on the section `Contributors Past & Present:` subsection `Individuals (in alphabetical order)`

File: doc/src/pgRouting-introduction.rst [hunk 10]
```diff
@@ -68,6 +68,7 @@ Individuals in this release v4.0.0 (in alphabetical order)
 
 Bipasha Gayary,
 Fan Wu,
+Mohit Rawat,
```
Comment (cvvergara) [inline doc/src/pgRouting-introduction.rst:69]: Probably you didnt read the [comment](https://github.com/pgRouting/pgrouting/pull/3019#issuecomment-3708324068) of the change in this lines

Please make the whole fix

File: pgtap/withPoints/withPoints/edge_cases/issue_2966.pg [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  END IF;
+
+  CREATE TEMP TABLE test_edges AS
+  SELECT * FROM (VALUES
+    (100::BIGINT, 1001::BIGINT, 1002::BIGINT, 10::FLOAT, 10::FLOAT),
+    (101::BIGINT, 1002::BIGINT, 1003::BIGINT, 12::FLOAT, 12::FLOAT),
+    (102::BIGINT, 1002::BIGINT, 1003::BIGINT, 100::FLOAT, 100::FLOAT)
+  ) AS t(id, source, target, cost, reverse_cost);
+
+  -- Test 1: fraction = 0.99 should work correctly (baseline)
```
Comment (cvvergara) [inline pgtap/withPoints/withPoints/edge_cases/issue_2966.pg]: You forgot to fix the comments and the message on this file

File: pgtap/withPoints/withPointsVia/edge_cases/issue_2966.pg [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  END IF;
+
+  CREATE TEMP TABLE test_edges AS
+  SELECT * FROM (VALUES
+    (100::BIGINT, 1001::BIGINT, 1002::BIGINT, 10::FLOAT, 10::FLOAT),
+    (101::BIGINT, 1002::BIGINT, 1003::BIGINT, 12::FLOAT, 12::FLOAT),
+    (102::BIGINT, 1002::BIGINT, 1003::BIGINT, 100::FLOAT, 100::FLOAT)
+  ) AS t(id, source, target, cost, reverse_cost);
+
+  -- Test 1: fraction = 1 should work correctly
```
Comment (cvvergara) [inline pgtap/withPoints/withPointsVia/edge_cases/issue_2966.pg]: You forgot to fix the comments and message on this other file

File: src/withPoints/withPoints.cpp [hunk 13]
```diff
@@ -397,8 +397,39 @@ Pg_points_graph::create_new_edges() {
             if (point.fraction == 1) {
                 log << "point's vertex_id = target" << edge.target << "\n";
                 point.vertex_id = -point.pid;
-                Edge_t new_edge = {edge.id, edge.target, -point.pid, 0, 0};
-                m_new_edges.push_back(new_edge);
+                // Zero-cost edge connecting target node to point (they're at same location)
```
Comment (cvvergara) [inline src/withPoints/withPoints.cpp]: Just noticed:
Please change to C++ comment style
```
/* comment */
```

And if you see more comments with C style on this file change them please

File: src/withPoints/withPoints.cpp [hunk 14]
```diff
@@ -397,8 +397,39 @@ Pg_points_graph::create_new_edges() {
...
+                Edge_t zero_edge = {edge.id, edge.target, -point.pid, 0, 0};
+                m_new_edges.push_back(zero_edge);
+
+                // Also create edge from previous target to this point
```
Comment (cvvergara) [inline src/withPoints/withPoints.cpp]: More than 2 line comments:
```
/*
   line 1
   line 2
*/
```

File: pgtap/withPoints/withPointsVia/edge_cases/issue_2966.pg [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  RETURN QUERY SELECT set_eq(
+    $$SELECT edge::TEXT FROM pgr_withPointsVia(
+      'SELECT * FROM test_edges',
+      'SELECT * FROM (VALUES (1, 100, 0.5), (2, 101, 1)) AS t(pid, edge_id, fraction)',
+      ARRAY[-1, -2]::BIGINT[],
+      'b'
+    ) WHERE edge > 0$$,
+    ARRAY['100', '101']::TEXT[],
+    'pgr_withPointsVia: Path should use edges 100 and 101'
+  );
```
[the remark this answers is not in the document]
Comment (cvvergara) [reply pgtap/withPoints/withPointsVia/edge_cases/issue_2966.pg:68]: Rabbit is correct, but we will leave those tests for another PR
